### ISSUE TITLE
regex: suggest reusing same pattern

### ISSUE THREAD

**Issue by epa** [Perl-Critic/Perl-Critic#1036]
The empty regexp reuses the last successful match.  This is useful in cases where you want to check for a match and then substitute.  A trivial example:
```
if (/oo/) {
    s/oo/uu/;
}
```
Perlcritic should suggest changing this to
```
if (/oo/) {
    s//uu/;
}
```

**Comment by petdance** [comment]
Thanks for the idea. I forgot about how Perl had those shortcuts.

In this case, I don't see why you would check before replacing.  This is functionally identical to

    if (s/oo/uu/) {
    }

Is there a reason we wouldn't suggest that?

**Comment by epa** [comment]
That code snippet is a trivial example.  I agree, you wouldn't actually write that.  But it does happen quite a lot in real code that you need to test a regexp first, then do something before replacing.

```
if (/foo/) {
   my $replacement = slow_and_expensive_call();
   s//$replacement/ or die;
}
```

**Comment by petdance** [comment]
OK, that's interesting.  You don't want to figure out the replacement unless you need to.  Hmmm.

The whole thing just feels fragile and easy to mess up. I'm not sure we want to encourage it. For example:

```
my $x = "whoopie";
my $y = "goop food book";

if ( $x =~ /oo/ ) {
    $y =~ s//XX/g;
}

say $x;
say $y;

whoopie
gXXp fXXd bXXk
```

Note that I checked against `$x` and replaced against `$y`.

Is there some advantage to not repeating the `oo` in the `s/oo/uu/`?

**Comment by petdance** [comment]
Here's a weird little coincidence.  Perl 5.38.0 now lets you get at that last successful match.

https://metacpan.org/release/RJBS/perl-5.38.0/view/pod/perldelta.pod

# New regexp variable ${^LAST_SUCCESSFUL_PATTERN}

This allows access to the last succesful pattern that matched in the current scope. Many aspects of the regex engine refer to the "last successful pattern". The empty pattern reuses it, and all of the magic regex vars relate to it. This allows access to its pattern. The following code

```
if (m/foo/ || m/bar/) {
    s//PQR/;
}
```
can be rewritten as follows

```
if (m/foo/ || m/bar/) {
    s/${^LAST_SUCCESSFUL_PATTERN}/PQR/;
}
```
and it will do the exactly same thing.

**Comment by epa** [comment]
Indeed, why do you think that after 25 years of Perl programming I only just now became aware of this shortcut...

**Comment by epa** [comment]
I agree that `if (/foo/) { s/foo/x/ }` can be an anti-pattern, and 99% of the time it's better to just do the replacement.  Nonetheless it's sometimes useful, which must be one reason why Perl supports the `//` shortcut.  One example might be
```
if (/foo=(\d+)/ && $1 > 50) {
    s//bigfoo/;
}
```
or in general, any case where you want to do some test (which is easier to express in Perl code than inside the regexp engine) before doing the substitution.

The regexp might not be as simple as `oo`.  You showed a possible mistake where the wrong variable was used (`$y` instead of `$x`) but it's just as likely, if a complex regexp were matched and then later replaced, that the two copies of the regexp would drift apart.  The `//` shortcut is not only more efficient but avoids repeating the same code.

**Comment by epa** [comment]
I agree that `if (/foo/) { s/foo/x/ }` can be an anti-pattern, and 99% of the time it's better to just do the replacement.  Nonetheless it's sometimes useful, which must be one reason why Perl supports the `//` shortcut.  One example might be
```
if (/foo=(\d+)/ && $1 > 50) {
    s//bigfoo/;
}
```
or in general, any case where you want to do some test (which is easier to express in Perl code than inside the regexp engine) before doing the substitution.

The regexp might not be as simple as `oo`.  You showed a possible mistake where the wrong variable was used (`$y` instead of `$x`) but it's just as likely, if a complex regexp were matched and then later replaced, that the two copies of the regexp would drift apart.  The `//` shortcut avoids repeating the same code.
